checks/libs/wmi/sampler.py

```diff
@@ -263,45 +259,32 @@ def _format_property_values(self, previous, current):
     @contextmanager
     def get_connection(self):
         """
-        Return an existing, available WMI connection or create a new one.
-        Release, i.e. mark as available at exit.
+        Create a new WMI connection
         """
-        connection = None
-
-        # Fetch an existing connection or create a new one
-        available_connections = self._wmi_connections[self.connection_key]
-
-        if available_connections:
-            self.logger.debug(
-                u"Using cached connection "
-                u"(host={host}, namespace={namespace}, username={username}).".format(
-                    host=self.host,
-                    namespace=self.namespace,
-                    username=self.username
-                )
-            )
-            connection = available_connections.pop()
-        else:
-            self.logger.debug(
-                u"Connecting to WMI server "
-                u"(host={host}, namespace={namespace}, username={username}).".format(
-                    host=self.host,
-                    namespace=self.namespace,
-                    username=self.username
-                )
-            )
-            locator = Dispatch("WbemScripting.SWbemLocator")
-            connection = locator.ConnectServer(
-                self.host, self.namespace,
-                self.username, self.password
+        self.logger.debug(
+            u"Connecting to WMI server "
+            u"(host={host}, namespace={namespace}, username={username}).".format(
+                host=self.host,
+                namespace=self.namespace,
+                username=self.username
             )
+        )
+
+        # Initialize COM for the current thread
+        # WARNING: any python COM object (locator, connection, etc) created in a thread
+        # shouldn't be used in other threads (can lead to memory/handle leaks if done
+        # without a deep knowledge of COM's threading model). Because of this and given
+        # that we run each query in its own thread, we don't cache connections
+        pythoncom.CoInitialize()
+        locator = Dispatch("WbemScripting.SWbemLocator")
+        connection = locator.ConnectServer(
+            self.host, self.namespace,
+            self.username, self.password
+        )
 
         # Yield the connection
         yield connection
 
```

Comment:
Blows my mind that holding a reference to the `connection` that would then get overwritten in subsequent runs wouldn't be garbage collected. Good catch.